frontend/lib/resolution/InitResolver.cpp

```diff
@@ -769,13 +769,16 @@ bool InitResolver::handleAssignmentToField(const OpCall* node) {
       // field doesn't contribute to the receiver type.
       updateResolverVisibleReceiverType();
 
-      auto lhsKind = state->qt.kind();
-      if (lhsKind != QualifiedType::TYPE && lhsKind != QualifiedType::PARAM) {
+      auto& reLhs = initResolver_.byPostorder.byId(lhs->id());
```

Comment:
Notably (see above) the state `qt` is still determined from `reRhs` type, which I don't think is right. Later on in the process (see `computeReceiverTypeConsideringState`, the QT of the state is used for constructing instantiations explicitly. I don't think this is right, since in this particular case, this will lead to using `null` (the type) as the field type.

Does call-init-deinit re-run field resolution with results of `init=`? If not, I think this will lead to incorrect instantiations.

To see this, try with a generic class field.

Reply:
I don't think call-init-deinit reruns field resolution; it looks like it assumes the field types have been resolved already. 

Could you elaborate on why a generic class field would expose this behavior? I think there's holes in my understanding of the process for generics here.

Reply:
Well, there's a bunch of logic that computes the field types / field substitutions for a generic class when it's being constructed. E.g., so that:

```Chapel
proc init() {
  self.myTypeField = int;
}
```

Leads to a type where `myTypeField` has an `int` substitution. This is only necessary for generic fields (because concrete fields don't need instantiations). This is why I'm telling you to try with a generic field (and it needs to be a class field so that `nil` can be legally assigned to it). 

https://github.com/chapel-lang/chapel/blob/939475fa0defad3f741d43f18b3913afbb05abbb/frontend/lib/resolution/InitResolver.cpp#L361

When inserting the substitutions, it looks like the type of `state->qt` is used directly. 

https://github.com/chapel-lang/chapel/blob/939475fa0defad3f741d43f18b3913afbb05abbb/frontend/lib/resolution/InitResolver.cpp#L364-L365

This means that, since your PR doesn't change `state->qt` from its original (RHS) value, a generic field being instantiated using `nil` will get the type of `nil`. I don't think that is what you want.

Reply:
The thing I'm concretely telling you to try is something like the following:

```
class C {
  type typeField; // C is now generic
}

record R {
  var myC: owned C(?)?;

  proc init() {
    this.myC = nil; // here, state->qt for myC is NullType, so you'll get R(NullType) instead of R(C(...))
  }
}
```